### PR TITLE
Change most of the Deserializer to use `forward_to_deserialize_any`

### LINKED ISSUE #27: Crate panics on incomplete struct deserialization
I'd personally prefer to replace both todos with NOPs, or at least have them be an error. I will say when working with the binary RPC there's truly no value to requiring complete structs and if it's meant for people doing further work to be able to ensure their data is complete, serde provides `deny_unknown_fields` (yet not an allow_unknown_fields which allows bypassing these todos).

### PR DESCRIPTION
**Pull request by Boog900**
fixes #27 

~~This also changes my old commit fixing structs with embedded structs to use a different method. A bool `read_header` instead of putting a struct marker at the front of the array of data.~~

This also fixes the panics on incomplete structs, just forwarding to forward_to_deserialize_any.

I could have removed more of the de-serialize but I feel the explicit errors on unsupported types may be helpful.

I have also changed how the Deserializer handles string markers, this is because epee marks byte arrays as strings so if someone sends a bytes array in a field we don't have this will get forwarded to deserialize_any which will then fail if the byte array isn't an actual string.

@thomaseizinger

### CODE REVIEW

File: src/de.rs [hunk 1]
```diff
@@ -50,11 +54,13 @@ impl<'b> Deserializer<'b> {
         Ok(value)
     }
 
-    fn read_varint_string(&mut self) -> Result<String> {
-        let string_length = self.read_varint()?;
-        let string = self.read_string(string_length)?;
-
-        Ok(string)
+    fn read_varint_marked_string(&mut self) -> Result<Vec<u8>> {
```
Comment (thomaseizinger) [inline src/de.rs:57]: Why does this say `string` when it returns `Vec<u8>`?

File: src/de.rs [hunk 2]
```diff
@@ -262,80 +268,19 @@ impl<'de, 'a, 'b> serde::Deserializer<'de> for &'a mut Deserializer<'b> {
...
-
-    fn deserialize_u32<V>(self, visitor: V) -> Result<<V as Visitor<'de>>::Value>
-    where
-        V: Visitor<'de>,
-    {
-        self.read_expected_marker(MARKER_SINGLE_U32)?;
-        self.dispatch_based_on_marker(MARKER_SINGLE_U32, visitor)
+        if !self.read_header {
+            self.read_header = true;
+            visitor.visit_map(MapAccess::with_varint_encoded_fields(self)?)
```
Comment (thomaseizinger) [inline src/de.rs:273]: This change seems to be orthogonal, correct? Can we have that in a separate PR please?

**Review by thomaseizinger — COMMENTED** [review]
Thanks!

Two comments but overall ACK.

File: src/de.rs (hunk 1 above)
Comment (Boog900) [reply src/de.rs:57]: This function decodes the epee field which is marked as a string but yeah that is a bit confusing so I'll change it

**Comment by Boog900** [comment]
I have opened another PR for changing how embedded structs are handled
